src/main/scala/edu/berkeley/ce/sparkrocks/Block.scala

```diff
@@ -532,41 +571,70 @@ case class Block(center: Array[Double], faces: Seq[Face], generation: Int=0) ext
-        // Factor of 3 comes from: centroid / (2.0 * (volume/6.0))
-        3.0 * centroidX / totalVolume,
-        3.0 * centroidY / totalVolume,
-        3.0 * centroidZ / totalVolume
-      )
+      val increments = faces.flatMap { face =>
+        // First vertex of face used in all iterations
+        val anchorVertex = vertices(face)(0)
+
+        // Tetrahedra are constructed using three vertices at a time
+        vertices(face).takeRight(vertices(face).length - 1).sliding(2).map { faceVertices =>
+          // Translate vertex coordinates from global to local where center of mass is local origin
+          val vertex1 = Array[Double](anchorVertex(0) - centerX,
+            anchorVertex(1) - centerY,
+            anchorVertex(2) - centerZ)
+          val vertex2 = Array[Double](faceVertices(0)(0) - centerX,
+            faceVertices(0)(1) - centerY,
+            faceVertices(0)(2) - centerZ)
+          val vertex3 = Array[Double](faceVertices(1)(0) - centerX,
+            faceVertices(1)(1) - centerY,
+            faceVertices(1)(2) - centerZ)
+
+          // Initialize Jacobian
+          val Jacobian = DenseMatrix(vertex1, vertex2, vertex3)
+
+          val JacobianDet = linalg.det(Jacobian)
+
+          // Calculate x, y and z centroid increments
+          ( (vertex1(0) + vertex2(0) + vertex3(0)) * JacobianDet,
+            (vertex1(1) + vertex2(1) + vertex3(1)) * JacobianDet,
+            (vertex1(2) + vertex2(2) + vertex3(2)) * JacobianDet
+            )
+        }
+      }
+
+      val (centroidX, centroidY, centroidZ) = increments.fold (0.0, 0.0, 0.0) {
+        case ((centX1, centY1, centZ1),  (centX2, centY2, centZ2)) =>
+          (centX1 + centX2, centY1 + centY2, centZ1 + centZ2)
+      }
+
+      Array(centroidX / (volume * 24.0) + centerX,
+        centroidY / (volume * 24.0) + centerY,
+        centroidZ / (volume * 24.0) + centerZ)
     }
   }
 
   /**
-    * Calculates the volume of the block
-    * See http://wwwf.imperial.ac.uk/~rn/centroid.pdf for theoretical background.
+    * Calculates the volume of the block using simplex integration
     *
     * @return The volume of the block
     */
   private def findVolume: Double = {
-    val vertices = findVertices
-    val mesh = meshFaces(vertices)
+    val vertices = orientedVertices
 
     val volIncrements = faces.flatMap { face =>
-      mesh(face).map { m =>
-        // Access triangulation entries backwards so they are in counterclockwise order
-        val a_vals = vertices(face)(m(2))
-        val b_vals = vertices(face)(m(1))
-        val c_vals = vertices(face)(m(0))
-        val a = DenseVector[Double](a_vals(0), a_vals(1), a_vals(2))
-        val b = DenseVector[Double](b_vals(0), b_vals(1), b_vals(2))
-        val c = DenseVector[Double](c_vals(0), c_vals(1), c_vals(2))
-
-        val n = linalg.cross(b - a, c - a)
-        a dot n
+      // Tetrahedra are constructed using three vertices at a time
+      vertices(face).sliding(3).map { vertexTuples =>
```

Comment:
How about doing `map { case (vertex1, vertex2, vertex3) => }` to avoid writing `vertexTuples(0)` and so on? Do you think that would simplify the code?